src/gpt.py

```diff
@@ -0,0 +1,312 @@
+    return user.name
+
+
+# Get totally random triggers about how to act:
+def get_random_context():
+    context = []
+    value = random.random()
+    if value < 0.25:
+        context.append("Respond in 10 words or less")
+    elif value < 0.5:
+        context.append("Be succinct")
+    elif value < 0.75:
+        context.append("Don't be too wordy")
+    else:
+        context.append("Write long detailed responses")
+    return context
+
+
+# Get context about the server
+async def get_server_context(message: Message) -> str:
+    result = []
+    # Load server event info
+    if message.guild and message.guild.scheduled_events:
+        # No idea if these are sorted by time
+        next_event = message.guild.scheduled_events[0]
+        if ":" in next_event.name:
+            event_num, event_topic = next_event.name.split(":", maxsplit=1)
+            result.append(f"Next bust event: {event_num.strip()}")
+            result.append(f"Next bust topic: {event_topic.strip()}")
+        else:
+            result.append(f"Next bust event: {next_event.name}")
+
+        result.append(f"Next bust time: {next_event.start_time.strftime('%b %d')}")
+        result.append(f"Today's date: {datetime.date.today().strftime('%b %d')}")
+
+    user = get_name(message.author)
+
+    # Role-based info
+    if isinstance(message.author, Member):
+        roles = {role.name for role in message.author.roles}
+
+        # Provide champion info
+        champ = [
+            ("Defending Champion", "first"),
+            ("Runner-up", "second"),
+            ("Bronzer", "third"),
+        ]
+        for role, place in champ:
+            if role in roles:
+                result.append(f"{user}'s place last bust: {place}")
+                break
+
+    result.append(f"Talking to: {user}")
+    return result
+
+
+# Count tokens in a string
+def token_count(data: str) -> int:
+    return len(encoding.encode(data))
+
+
+# Replace Discord-style mentions with names
+def strip_mentions(message: Message) -> str:
+    content = message.content
+    for user in message.mentions:
+        content = content.replace(user.mention, get_name(user))
+    for channel in message.channel_mentions:
+        content = content.replace(channel.mention, channel.name)
+    for role in message.role_mentions:
+        content = content.replace(role.mention, role.name)
+    return content
+
+
+# Fetch messages from history up to a certain token allowance
+async def fetch_history(
+    token_limit: int, speaking_turn_limit: int, message: Message
+) -> List[Tuple[str, bool]]:
+    total_tokens = 0
+    history = []
+    idx = 0
+    one_hour = datetime.timedelta(hours=1)
+    seen_message = False
+    speaking_turn_count = 0
+    last_author = None
+    async for msg in message.channel.history():
+        # Skip messages newer than what we're replying to
+        seen_message = seen_message or (msg == message)
+        if not seen_message:
+            continue
```

Comment:
Trying this and it seems a little strange. The API makes it sound like you can pass in either an `int` id or a `datetime`, but actually you pass in an object which has an "id" attribute (aka the message itself) or a datetime. Using `before=message` doesn't quite work for our code flow since it doesn't include `message` itself. I tried using `before=message.created_at+datetime.timedelta(microseconds=1)` but even that seems to not work so well. In fact even with a timedelta of one second it seems to not work?

There might be a way to "zip" together [message] + message.channel.history(before=message) in an async iterator friendly way (aka without flattening in advance) but I'm not sure if it would add too much complication to be worth it

Reply:
Oh yeah, I had misread [`nextcord.abc.Snowflake`](https://docs.nextcord.dev/en/stable/api.html#nextcord.abc.Snowflake) as an alias of `int`, rather than having an `int` attribute.

Indeed, it looks like it's going to be more complicated to reason about prepending to nextcord's `AsyncIterator` than just skipping via this way, especially since 99% of the time we'll be responding to the latest message in history anyways.